torch/csrc/jit/passes/peephole_list_idioms.cpp

```diff
@@ -21,17 +22,312 @@ c10::optional<size_t> normalizeIndex(int64_t index, size_t len) {
   }
 }
 
+// Refine from Value -> len of list
+// TODO: vector may be faster
+using ListRefinement = std::unordered_map<Value*, int64_t>;
+
+// The intersection of the refinements is the Value* which are in both
+// refinements and are refined to the same length
+ListRefinement intersectRefinements(
+    const ListRefinement& ref1,
+    const ListRefinement& ref2) {
+  ListRefinement out;
+  for (const auto& pair : ref1) {
+    auto val2 = ref2.find(pair.first);
+    if (val2 != ref2.end() && val2->second == pair.second) {
+      out[pair.first] = pair.second;
+    }
+  }
+  return out;
+}
+
+// To union, just take all refinements from both inputs. We do not need to worry
+// about len refinements disagreeing because a path like `if len(x) == 4 and
+// len(x) == 5` will never be taken
```

Comment:
Adding an example on when UnionRefinements are used would be helpful for understanding.